### PR TITLE
Reduce memory usage of the HTTP transport

### PR DESCRIPTION
**Pull request by marcotc**
_**tl;dr: 21-50% memory reduction, 19-30% faster HTTP transport.**_

This PR reduces memory of our default HTTP transport. As result of these changes, performance has been increased as well.

The largest gains were due:
* Direct MessagePack serialization of `span.rb`: before we were creating an intermediate `Hash` object in order to serialize spans; now we directly interface with MessagePack objects.
* `Net::HTTP.start` is [inefficient when processing its named arguments](https://github.com/ruby/ruby/blob/b2d96abb42abbe2e01f010ffc9ac51f0f9a50002/lib/net/http.rb#L614-L618): we now use a different interface to configure these options, which avoids the expensive processing step.

One area of with large possible improvements is to use a different HTTP adapter. The native `Net::HTTP` is great because it's always available, but it has showed up many times in the memory profiler, due to many strings being created during the processing of HTTP requests and responses.

## Results
`spec/ddtrace/benchmark/transport_benchmark_spec.rb` includes the benchmarks used to profile the tracer and produce the numbers reported here.

### Memory
```
Before [bytes allocated (objects created)]
   1  span:   3601720   (53508)
  10 spans:   5503620   (63408)
 100 spans:  24437620  (162408)
1000 spans: 214035300 (1152590)

After [bytes allocated (objects created)]
   1  span:   2838120   (47208)
  10 spans:   3782420   (54408)
 100 spans:  13143620  (126408)
1000 spans: 106960100  (846590)

Difference (% reduction)
   1  span: 21% (11%)
  10 spans: 31% (14%)
 100 spans: 46% (22%)
1000 spans: 50% (26%)
```

### CPU
```
Before [operations/sec]
     1 spans:    1976.05
    10 spans:    1820.21
   100 spans:     918.89
  1000 spans:     147.81

After [operations/sec]
     1 spans:    2576.38
    10 spans:    2316.30
   100 spans:    1177.31
  1000 spans:     175.92

Comparison (% faster; slower if negative)
     1 spans:      30.38
    10 spans:      27.25
   100 spans:      28.12
  1000 spans:      19.02
```

### CODE REVIEW

File: lib/ddtrace/span.rb [hunk 1]
```diff
@@ -254,6 +254,61 @@ def to_hash
...
 
+    # MessagePack serializer interface. Making this object
+    # respond to `#to_msgpack` allows it to be automatically
+    # serialized by MessagePack.
+    #
+    # This is more efficient than doing +MessagePack.pack(span.to_hash)+
+    # as we don't have to create an intermediate Hash.
+    #
+    # @param packer [MessagePack::Packer] serialization buffer, can be +nil+ with JRuby
+    def to_msgpack(packer = nil)
```
Comment (brettlangdon) [inline lib/ddtrace/span.rb:265]: @Kyle-Verhoog can you review this?

Are there any optimizations you found from the Python implementation that we might be able to borrow here as well?

File: spec/ddtrace/benchmark/transport_benchmark_spec.rb [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+require_relative 'support/benchmark_helper'
+
+require 'socket'
+
+RSpec.describe 'Microbenchmark Transport' do
+  context 'with HTTP transport' do
+    # Create server that responds just like the agent,
+    # but doesn't consume as many resources, nor introduces external
+    # noise into the benchmark.
+    let(:server) { TCPServer.new '127.0.0.1', ENV[Datadog::Ext::Transport::HTTP::ENV_DEFAULT_PORT].to_i }
```
Comment (brettlangdon) [inline spec/ddtrace/benchmark/transport_benchmark_spec.rb:12]: How does TCPServer work? Does it run in a separate process?

**Review by brettlangdon — COMMENTED** [review]
Should we separate our the changes of direct msgpack encoding from the transport changes?

Can we add benchmark suite for the msgpack changes on it's own? (e.g. encoding a trace vs using the transport)

File: spec/ddtrace/benchmark/transport_benchmark_spec.rb (hunk 2 above)
Comment (marcotc) [reply spec/ddtrace/benchmark/transport_benchmark_spec.rb:12]: It runs on the same process, but separate thread. Any specific reason for asking?

File: lib/ddtrace/span.rb [hunk 3]
```diff
@@ -254,6 +254,61 @@ def to_hash
...
+        packer.write(:start)
+        packer.write((@start_time.to_f * 1e9).to_i)
+
+        packer.write(:duration)
+        packer.write(((@end_time - @start_time) * 1e9).to_i)
+      else
+        packer.write_map_header(11) # Set header with how many elements in the map
+      end
+
+      packer.write(:span_id)
```
[the remark this answers is not in the document]
Comment (marcotc) [reply lib/ddtrace/span.rb:281]: They are converted to string by MessagePack, but Ruby has an internalized string for each symbol.
Although I think it's worth trying it with strings directly, as there's a chance Ruby could create a copy of that internalized string each time we ask for it.

I'll report on results here, thanks for the heads up!

[the remark this answers is not in the document]
Comment (marcotc) [reply lib/ddtrace/span.rb]: It did improve the results, thank you @Kyle-Verhoog! 🎉 
Memory usage stayed the same, as symbols already have an internal string representation, but performance was improved.

File: spec/ddtrace/benchmark/transport_benchmark_spec.rb (hunk 2 above)
Comment (brettlangdon) [reply spec/ddtrace/benchmark/transport_benchmark_spec.rb:12]: make sense, my reasoning for asking was, introducing something like this in-process, will it impact performance of results measured due to the GIL?

probably not... ? since it all synchronous anyways?

File: lib/ddtrace/span.rb [hunk 4]
```diff
@@ -269,36 +271,36 @@ def to_msgpack(packer = nil)
       if !@start_time.nil? && !@end_time.nil?
         packer.write_map_header(13) # Set header with how many elements in the map
 
-        packer.write(:start)
+        packer.write('start')
```
Comment (brettlangdon) [inline lib/ddtrace/span.rb:274]: worth adding a comment on why we use strings instead of symbols?

any benefits to making these constants/freezing them outside the scope of this method?

**Comment by brettlangdon** [comment]
I thought I had a comment somewhere, but don't see it. Do we have benchmarks specifically for the encoding piece?

e.g. `to_msgpack` of a single span, or a trace of varying sizes, outside the context of the transport?

File: lib/ddtrace/span.rb (hunk 4 above)
Comment (marcotc) [reply lib/ddtrace/span.rb:274]: > any benefits to making these constants/freezing them outside the scope of this method?

Instead of doing that, given there are so many strings,  I added `# frozen_string_literal: true` to the top of the file, which freezes all strings in this. I did check all strings declared in this file and they are all safe to freeze.

I also benchmark only the `# frozen_string_literal: true` change, to see if the other strings frozen in the file would make a change to our numbers, but they didn't, so the performance improvement does come from the change from symbol to string only.

> worth adding a comment on why we use strings instead of symbols?

I'll add a comment for that, good call!

**Comment by marcotc** [comment]
> e.g. to_msgpack of a single span, or a trace of varying sizes, outside the context of the transport?

We could add these, and ultimately also test our JSON serialization.
Where we stand today, the transport benchmark in PR is pretty much 50% serialization, 50% HTTP client.

I think we have it pretty well covered at this moment, but we can add this more granular benchmark as well.

File: spec/ddtrace/benchmark/transport_benchmark_spec.rb (hunk 2 above)
Comment (marcotc) [reply spec/ddtrace/benchmark/transport_benchmark_spec.rb:12]: I tried with forking. The difference is small, but I think that not boggling the GVL, even if not by very long, is probably the best way to go here.

For JRuby, which doesn't support forking, we still use threads, as it supports true parallel execution in threads.

**Comment by brettlangdon** [comment]
More granular might make sense as we can use it to optimize that specific piece.

e.g. if we want to improve `to_msgpack` also testing the http piece at the same time might make it noisy.

File: spec/ddtrace/benchmark/transport_benchmark_spec.rb (hunk 2 above)
Comment (brettlangdon) [reply spec/ddtrace/benchmark/transport_benchmark_spec.rb:12]: 👍🏻

**Comment by marcotc** [comment]
@brettlangdon cool, I scheduled a separate follow up task to benchmark specifically the serialization.

**Comment by brettlangdon** [comment]
@marcotc that sounds good to me, thanks for tracking it!

**Comment by brettlangdon** [comment]
Should we separate the span encoding changes from the transport changes?

e.g. add `to_msgpack` and the microbenchmarks there in a separate PR?

File: lib/ddtrace/span.rb [hunk 5]
```diff
@@ -254,6 +256,64 @@ def to_hash
...
+    # @param packer [MessagePack::Packer] serialization buffer, can be +nil+ with JRuby
+    def to_msgpack(packer = nil)
+      # As of 1.3.3, JRuby implementation doesn't pass an existing packer
+      packer ||= MessagePack::Packer.new
+
+      if !@start_time.nil? && !@end_time.nil?
+        packer.write_map_header(13) # Set header with how many elements in the map
+
+        packer.write('start')
+        packer.write((@start_time.to_f * 1e9).to_i)
```
Comment (brettlangdon) [inline lib/ddtrace/span.rb:275]: Since this is written in two places now, read only property?

`#start_time_ns` + `#duration_ns` ?

**Comment by marcotc** [comment]
@brettlangdon I don't think we need to separate it. At the end of day, encoding is an integral part of the "please send these spans to Datadog" part of our tracer, which is what we are benchmarking here.

**Review by brettlangdon — COMMENTED** [review]
What versions of Ruby have we benchmarked these changes with?

I see there is a case to use `filter_map` on >= 2.7, what kind of performance difference is there from this in 2.6 vs 2.7 (same for the other pieces)?

File: lib/ddtrace/span.rb (hunk 5 above)
Comment (marcotc) [reply lib/ddtrace/span.rb]: Cool, updated it.
I also changed the operation to perform fewer operations to get the value as nanoseconds.
The move to a separate method, combined with the arithmetic improvements increased performance bit a very tiny bit.

**Comment by marcotc** [comment]
@brettlangdon Results are for the fastest version, 2.7.
The difference between 2.7 and 2.6, for example, is very small. The change to `filter_map` reduced memory usage bit a small amount and increased performance by a very small margin. Considering the whole PR, this is sub-percent improvement.

File: lib/ddtrace/transport/http/adapters/net.rb [hunk 6]
```diff
@@ -20,8 +20,14 @@ def initialize(hostname, port, options = {})
           end
 
           def open
-            # Open connection
-            ::Net::HTTP.start(hostname, port, open_timeout: timeout, read_timeout: timeout) do |http|
+            # DEV Initializing +Net::HTTP+ directly help us avoid expensive
+            # options processing done in +Net::HTTP.start+:
+            # https://github.com/ruby/ruby/blob/b2d96abb42abbe2e01f010ffc9ac51f0f9a50002/lib/net/http.rb#L614-L618
+            req = ::Net::HTTP.new(hostname, port, nil)
```
Comment (ericmustin) [inline lib/ddtrace/transport/http/adapters/net.rb:26]: I know this is kinda out there but have we considered using a 3rd party http library? Might not be worth the pain but I believe some other vendors use [`http.rb`](https://github.com/honeycombio/libhoney-rb/blob/00dd785ce796c193db0a628fd4be251a2353fe69/libhoney.gemspec#L37)

File: lib/ddtrace/span.rb [hunk 7]
```diff
@@ -247,13 +249,71 @@ def to_hash
...
+
+        packer.write('duration')
+        packer.write(duration_nano)
+      else
+        packer.write_map_header(11) # Set header with how many elements in the map
+      end
+
+      # DEV: We use strings as keys here, instead of symbols, as
+      # DEV: MessagePack will ultimately convert them to strings.
+      # DEV: By providing strings directly, we skip this indirection operation.
```
Comment (ericmustin) [inline lib/ddtrace/span.rb:285]: This would be faster than defining a bunch of constants, ie 

```
SPAN_ID = 'span_id'.freeze
...
...
...
packer.write(SPAN_ID)
```
etc etc ?

File: lib/ddtrace/transport/traces.rb [hunk 8]
```diff
@@ -57,7 +57,12 @@ def initialize(encoder, max_size: DEFAULT_MAX_PAYLOAD_SIZE)
         # @return [Enumerable[Array[Bytes,Integer]]] list of encoded chunks: each containing a byte array and
         #   number of traces
         def encode_in_chunks(traces)
-          encoded_traces = traces.map { |t| encode_one(t) }.reject(&:nil?)
+          encoded_traces = if traces.respond_to?(:filter_map)
+                             # DEV Supported since Ruby 2.7, saves an intermediate object creation
+                             traces.filter_map { |t| encode_one(t) }
```
Comment (ericmustin) [inline lib/ddtrace/transport/traces.rb:62]: 👍

File: lib/ddtrace/span.rb (hunk 7 above)
Comment (ericmustin) [reply lib/ddtrace/span.rb:285]: nvmd, i see discussion here: https://github.com/DataDog/dd-trace-rb/pull/1165#discussion_r483825973

**Review by ericmustin — APPROVED** [review]
Lgtm, the only real speed i could think of would be tinkering with the http client itself, but probably not worth the risks involved of swapping in a 3rd party library, and i guess would also increase memory usage of the tracer on startup

File: lib/ddtrace/transport/http/adapters/net.rb (hunk 6 above)
Comment (marcotc) [reply lib/ddtrace/transport/http/adapters/net.rb:26]: We have a follow up task to investigate this 👍

File: lib/ddtrace/span.rb [hunk 9]
```diff
@@ -303,5 +363,17 @@ def now_allocations
...
+    # Used for serialization
+    # @return [Integer] in nanoseconds since Epoch
+    def start_time_nano
+      @start_time.to_i * 1000000000 + @start_time.nsec
+    end
+
+    # Used for serialization
+    # @return [Integer] in nanoseconds since Epoch
+    def duration_nano
+      ((@end_time - @start_time) * 1e9).to_i
```
[the remark this answers is not in the document]
Comment (marcotc) [reply lib/ddtrace/span.rb:376]: 👍

**Comment by marcotc** [comment]
After rebasing changes to use monotonic clock, no visible performance impact can be measured.

**Review by ericmustin — APPROVED** [review]
🚀

**Comment by marcotc** [comment]
@ericmustin I forgot I based #1178 on top of this branch (because of some shared fixtures that it easier to write future benchmarks). Would you mind ✅  this again when you have some time?

**Review by ericmustin — APPROVED** [review]
